src/main/java/io/vertx/core/net/impl/HostAndPortImpl.java

```diff
@@ -30,35 +30,45 @@ public static int parseIPv4Address(String s, int from, int to) {
         return -1;
       }
     }
-    return from < to && (from + 1 == s.length() || s.charAt(from + 1) != ':') ? -1 : from;
+    // from is the next position to parse: whatever come before is a valid IPv4 address
+    if (from == to) {
+      // we're done
+      return from;
+    }
+    assert from < to;
+    // we have more characters, let's check if it has enough space for a port
+    if (from + 1 == s.length()) {
+      // just a single character left, we don't care what it is
+      return -1;
+    }
+    // we have more characters
+    if (s.charAt(from) != ':') {
+      // we need : to start a port
+      return -1;
+    }
+    // we (maybe) have a port - even with a single digit; the ipv4 addr is fineFi
+    return from;
   }
 
   static int parseDecOctet(String s, int from, int to) {
     int val = parseDigit(s, from++, to);
-    switch (val) {
```

Comment:
we have the exact boundaries here, I don't see the point to use a switch - at the mercy of the JIT to improve it